unstructured/ingest/v2/processes/connectors/opensearch.py

```diff
@@ -0,0 +1,118 @@
+from typing import TYPE_CHECKING, Any, Optional
+
+from unstructured.ingest.enhanced_dataclass import enhanced_field
+from unstructured.ingest.error import DestinationConnectionError
+from unstructured.ingest.utils.data_prep import generator_batching_wbytes
+from unstructured.ingest.v2.interfaces import (
+    AccessConfig,
+    ConnectionConfig,
+    UploadContent,
+    Uploader,
+)
+from unstructured.ingest.v2.logger import logger
+from unstructured.ingest.v2.processes.connector_registry import (
+    DestinationRegistryEntry,
+    add_destination_entry,
+)
+from unstructured.ingest.v2.processes.connectors.elasticsearch import (
+    ElasticsearchUploaderConfig,
+    ElasticsearchUploadStager,
+    ElasticsearchUploadStagerConfig,
+)
+from unstructured.utils import requires_dependencies
+
+if TYPE_CHECKING:
+    from opensearchpy import OpenSearch
+
+CONNECTOR_TYPE = "opensearch"
+
+"""Since the actual OpenSearch project is a fork of Elasticsearch, we are relying
+heavily on the Elasticsearch connector code, inheriting the functionality as much as possible."""
+
+
+@dataclass
+class OpenSearchAccessConfig(AccessConfig):
+    password: Optional[str] = enhanced_field(default=None, sensitive=True)
+    use_ssl: bool = False
+    verify_certs: bool = False
+    ssl_show_warn: bool = False
+    ca_certs: Optional[str] = None
+    client_cert: Optional[str] = None
+    client_key: Optional[str] = None
+
+
+@dataclass
+class OpenSearchConnectionConfig(ConnectionConfig):
+    hosts: Optional[list[str]] = None
+    username: Optional[str] = None
+    access_config: OpenSearchAccessConfig = enhanced_field(sensitive=True)
+
+    def to_dict(self, **kwargs) -> dict[str, json]:
+        d = super().to_dict(**kwargs)
+        d.update(self.access_config.to_dict())
+        d["http_auth"] = (self.username, self.access_config.password)
+        return d
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["opensearchpy"], extras="opensearch")
+    def get_client(self) -> "OpenSearch":
+        from opensearchpy import OpenSearch
+
+        return OpenSearch(**self.to_dict(apply_name_overload=False))
+
+
+@dataclass
+class OpenSearchUploader(Uploader):
+    upload_config: ElasticsearchUploaderConfig
+    connection_config: OpenSearchConnectionConfig
+
+    def run(self, contents: list[UploadContent], **kwargs: Any) -> None:
+        elements_dict = []
+        for content in contents:
+            with open(content.path) as elements_file:
+                elements = json.load(elements_file)
+                elements_dict.extend(elements)
+        logger.info(
+            f"writing document batches to destination"
+            f" index named {self.upload_config.index_name}"
+            f" at {self.connection_config.hosts}"
+            f" with batch size (in bytes) {self.upload_config.batch_size_bytes}"
+            f" with {self.upload_config.thread_count} (number of) threads"
+        )
+        from opensearchpy.helpers import parallel_bulk
```

Comment:
Given how much overlap this has with the elasticsearch connector, should we find a way to extract how the method is fetched (from the lasticsaerch sdk or the opensearch one) and have this one inherit from the elasticsearch connector?

Reply:
We definitely could. The only holdback is due to the logging being misleading if it says Elasticsearch. 

Maybe we remove the `elasticsearch` word from the logging in the ES connector? That work? Then I can pretty much just inherit from ES.

Reply:
Or dynamically use the class name to populate that part of the log?

Reply:
Also want to call out given the logic itself is the same, I made some changes to the elasticsaerch connector logic on writing the data and by inheriting from that, we don't need to worry about copy-paste drift between the two.

Reply:
@rbiseck3 I shifted some things. It did affect the upload stager test.